### PR TITLE
Controller runtime v0.16

### PR DESCRIPTION
**Pull request by bshephar**

### CODE REVIEW

**Comment by bshephar** [comment]
No idea what this means:
```
/go/src/github.com/openstack-k8s-operators/operator/go.mod:5: unknown directive: toolchain
```

What is `openstack-k8s-operators/operator`

**Comment by bshephar** [comment]
Ahh, I see it. I'm inadvertently bumping the go version and toolchain is coming from here:
https://github.com/openstack-k8s-operators/infra-operator/pull/155/files#diff-1a135474df0925a580fa1c89d8925690426c80f754af7142363cb87b6d99ac70R3-R5

**Comment by bshephar** [comment]
Checking to see if the last commit here resolves it. If it does, I'll squash the last commit

**Comment by bshephar** [comment]
Ok, this error happens because we need to be running at least go 1.20. 
```
# sigs.k8s.io/controller-runtime/pkg/cache
/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.16.3/pkg/cache/cache.go:412:58: fields.Selector does not implement comparable
/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.16.3/pkg/cache/cache.go:442:20: fields.Selector does not implement comparable
note: module requires Go 1.20
```

If that's the case, I feel like we're better off just going straight to 1.21 as part of this effort.

**Comment by bshephar** [comment]
recheck

File: apis/go.mod [hunk 1]
```diff
@@ -1,36 +1,38 @@
 module github.com/openstack-k8s-operators/infra-operator/apis
 
-go 1.19
+go 1.21
+
+toolchain go1.21.1
```
Comment (gibizer) [inline apis/go.mod:5]: OK, this is a min constraint and go will use whatever that is newer and built into the available go 1.21.x installation

File: go.mod [hunk 2]
```diff
@@ -1,6 +1,6 @@
 module github.com/openstack-k8s-operators/infra-operator
 
-go 1.19
+go 1.21
```
Comment (gibizer) [inline go.mod:3]: lets make it consistent and either add go toolchain here or remove it from the apis/go.mod as well.

File: renovate.json
Comment (gibizer) [inline renovate.json:1]: we need to make an extra step after every service operator is bumped and remove the version pinning from https://github.com/openstack-k8s-operators/renovate-config/blob/main/default.json5

**Review by gibizer — COMMENTED** [review]
only a minor nit about toolchain, but the overall direction looks good.

File: apis/go.mod (hunk 1 above)
Comment (bshephar) [reply apis/go.mod:5]: I guess this is automatically added by something. But for consistency, I'll add it to the root go.mod too.

File: go.mod (hunk 2 above)
Comment (bshephar) [reply go.mod:3]: added toolchain here

**Review by gibizer — COMMENTED** [review]
Thanks. Looks good to me.

**Comment by bshephar** [comment]
Kuttl test issues here:
```
go: github.com/openstack-k8s-operators/openstack-operator imports
	github.com/openstack-k8s-operators/infra-operator/apis/memcached/v1beta1 tested by
	github.com/openstack-k8s-operators/infra-operator/apis/memcached/v1beta1.test imports
	sigs.k8s.io/controller-runtime/pkg/metrics/server: module sigs.k8s.io/controller-runtime@latest found (v0.17.0, replaced by sigs.k8s.io/controller-runtime@v0.14.6), but does not contain package sigs.k8s.io/controller-runtime/pkg/metrics/server
```

File: apis/memcached/v1beta1/webhook_suite_test.go [hunk 3]
```diff
@@ -31,22 +31,28 @@ import (
...
+	kscheme "k8s.io/client-go/kubernetes/scheme"
 	"k8s.io/client-go/rest"
 	ctrl "sigs.k8s.io/controller-runtime"
 	"sigs.k8s.io/controller-runtime/pkg/client"
 	"sigs.k8s.io/controller-runtime/pkg/envtest"
 	logf "sigs.k8s.io/controller-runtime/pkg/log"
 	"sigs.k8s.io/controller-runtime/pkg/log/zap"
+	"sigs.k8s.io/controller-runtime/pkg/webhook"
+
+	metricsserver "sigs.k8s.io/controller-runtime/pkg/metrics/server"
```
Comment (gibizer) [inline apis/memcached/v1beta1/webhook_suite_test.go:43]: the kuttl error referring to this line. But v0.17 has this package available https://github.com/kubernetes-sigs/controller-runtime/tree/v0.17.0/pkg/metrics/server . Based on the error message for some reason go is pulling the version of the controller runtime back from 0.17 to 0.14.6. The only dependency that I can think of doing this is lib-common: https://github.com/openstack-k8s-operators/lib-common/blob/main/modules/common/go.mod#L18 So I suggest to try to bump the runtime in lib-common first

**Comment by bshephar** [comment]
Yeah, so same error here:
```
go: github.com/openstack-k8s-operators/openstack-operator imports
	github.com/openstack-k8s-operators/infra-operator/apis/memcached/v1beta1 tested by
	github.com/openstack-k8s-operators/infra-operator/apis/memcached/v1beta1.test imports
	sigs.k8s.io/controller-runtime/pkg/metrics/server: module sigs.k8s.io/controller-runtime@latest found (v0.17.0, replaced by sigs.k8s.io/controller-runtime@v0.14.6), but does not contain package sigs.k8s.io/controller-runtime/pkg/metrics/server
```

I would say because we import openstack-operator, we need to also update openstack-operator. Which will mean a myriad of other dependencies need updating as well.

File: apis/memcached/v1beta1/webhook_suite_test.go (hunk 3 above)
Comment (bshephar) [reply apis/memcached/v1beta1/webhook_suite_test.go:43]: openstack-operator has a replaces for this:
https://github.com/openstack-k8s-operators/openstack-operator/blob/main/go.mod#L125

**Comment by bshephar** [comment]
Ok, those extra vars don't seem to have pushed this to Go 1.21:
~~~
go: errors parsing go.mod:\n/home/zuul/src/github.com/openstack-k8s-operators/infra-operator/go.mod:5: unknown directive: toolchain\nmake: ***
~~~

**Comment by bshephar** [comment]
Hmm, not able to reproduce that error locally:
```
[bshephar@fedora-work-848cd9c6bd-r9zjb infra-operator]$ make controller-gen
mkdir -p /home/bshephar/infra-operator/bin
test -s /home/bshephar/infra-operator/bin/controller-gen && /home/bshephar/infra-operator/bin/controller-gen --version | grep -q v0.11.1 || \
GOBIN=/home/bshephar/infra-operator/bin go install sigs.k8s.io/controller-tools/cmd/controller-gen@v0.11.1
go: downloading sigs.k8s.io/controller-tools v0.11.1
go: downloading github.com/spf13/cobra v1.6.1
go: downloading github.com/gobuffalo/flect v0.3.0
go: downloading k8s.io/apiextensions-apiserver v0.26.0
go: downloading k8s.io/apimachinery v0.26.0
go: downloading golang.org/x/tools v0.4.0
go: downloading gopkg.in/yaml.v2 v2.4.0
go: downloading github.com/fatih/color v1.13.0
go: downloading k8s.io/api v0.26.0
go: downloading gopkg.in/yaml.v3 v3.0.1
go: downloading sigs.k8s.io/yaml v1.3.0
go: downloading github.com/spf13/pflag v1.0.5
go: downloading github.com/gogo/protobuf v1.3.2
go: downloading k8s.io/utils v0.0.0-20221107191617-1a15be271d1d
go: downloading github.com/google/gofuzz v1.1.0
go: downloading k8s.io/klog/v2 v2.80.1
go: downloading sigs.k8s.io/structured-merge-diff/v4 v4.2.3
go: downloading github.com/mattn/go-colorable v0.1.9
go: downloading github.com/mattn/go-isatty v0.0.14
go: downloading golang.org/x/sys v0.3.0
go: downloading sigs.k8s.io/json v0.0.0-20220713155537-f223a00ba0e2
go: downloading gopkg.in/inf.v0 v0.9.1
go: downloading github.com/go-logr/logr v1.2.3
go: downloading golang.org/x/mod v0.7.0
go: downloading github.com/json-iterator/go v1.1.12
go: downloading golang.org/x/net v0.4.0
go: downloading github.com/modern-go/reflect2 v1.0.2
go: downloading github.com/modern-go/concurrent v0.0.0-20180306012644-bacd9c7ef1dd
go: downloading golang.org/x/text v0.5.0
[bshephar@fedora-work-848cd9c6bd-r9zjb infra-operator]$
[bshephar@fedora-work-848cd9c6bd-r9zjb infra-operator]$ go version
go version go1.21.6 linux/amd64
```

**Comment by bshephar** [comment]
Ah, yes I can. It's an issue with the lib-common PR:
```
[bshephar@fedora-work-848cd9c6bd-r9zjb infra-operator]$ make build
/home/bshephar/infra-operator/bin/controller-gen object:headerFile="hack/boilerplate.go.txt" paths="./..."
../../go/pkg/mod/github.com/openstack-k8s-operators/lib-common/modules/common@v0.3.1-0.20240129151020-c9467a8fbbfc/condition/funcs.go:26:2: k8s.io/api@v0.29.1: missing go.sum entry for go.mod file; to add it:
	go mod download k8s.io/api
../../go/pkg/mod/github.com/openstack-k8s-operators/lib-common/modules/common@v0.3.1-0.20240129151020-c9467a8fbbfc/condition/funcs.go:27:2: k8s.io/apimachinery@v0.29.1: missing go.sum entry for go.mod file; to add it:
	go mod download k8s.io/apimachinery
```

**Comment by bshephar** [comment]
Ok, moved past those issues. Now the problem is going to be hard to get passed, because the next thing that happens is that we replace the infra-operator in openstack-operator and try to deploy everything. So kuttl isn't going to pass here until we have made this change in each and every operator.

**Pull request merged**